# Assignment Week 2 

* Answer each question completely. 
* Include your code and summarize what you see in a brief narrative.

__Charting:__
Remember to make your charts as self-sufficient as possible (i.e., labels, titles, lack of clutter)  

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

df = pd.read_csv('https://raw.githubusercontent.com/msaricaumbc/DS_data/master/ds602/car_prices/car_prices.csv', low_memory=False)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 558837 entries, 0 to 558836
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   year          558837 non-null  int64  
 1   make          548536 non-null  object 
 2   model         548438 non-null  object 
 3   trim          548186 non-null  object 
 4   body          545642 non-null  object 
 5   transmission  493484 non-null  object 
 6   vin           558833 non-null  object 
 7   state         558837 non-null  object 
 8   condition     547043 non-null  object 
 9   odometer      558743 non-null  float64
 10  color         558088 non-null  object 
 11  interior      558088 non-null  object 
 12  seller        558837 non-null  object 
 13  mmr           558837 non-null  object 
 14  sellingprice  558837 non-null  int64  
 15  saledate      558837 non-null  object 
dtypes: float64(1), int64(2), object(13)
memory usage: 68.2+ MB


# Data Description:
- year: year the car was built (e.g., the model year)  
- make: the manufacturer. 
- model: specific model of the car  
- trim: subtype of the car model. 
- body: body style of the car. 
- transmission: type of transmission  
- state: state car is registered. 
- condition: numerical rating of the condition, higher is better.  
- odometer: number of miles the car has been driven.   
- color: color of the car.  
- interior: type of interior (e.g., seats). 
- mmr: estimated wholesale price of the car if auctioned.  
- sellingprice: actual sales price of the car.

## Question 1. Determine the number of missing values in each column.

In [2]:
df.isnull().sum()

year                0
make            10301
model           10399
trim            10651
body            13195
transmission    65353
vin                 4
state               0
condition       11794
odometer           94
color             749
interior          749
seller              0
mmr                 0
sellingprice        0
saledate            0
dtype: int64

Sample narrative description.

* isnull() checks for null values in each column and sum() function adds them up.
* Transmission coumn has most no.of null values with 65353
* vin has the least no. of null values with 4

## Question 2. Check the data types.
- Which columns are coded to the incorrect type? If any, correct them.

In [3]:
df.head()

year   make                model        trim   body transmission  \
0  2015    Kia              Sorento          LX    SUV    automatic   
1  2015    Kia              Sorento          LX    SUV    automatic   
2  2014    BMW             3 Series  328i SULEV  Sedan    automatic   
3  2015  Volvo                  S60          T5  Sedan    automatic   
4  2014    BMW  6 Series Gran Coupe        650i  Sedan    automatic   

                 vin state condition  odometer  color interior  \
0  5xyktca69fg566472    ca         5   16639.0  white    black   
1  5xyktca69fg561319    ca         5    9393.0  white    beige   
2  wba3c1c51ek116351    ca       4.5    1331.0   gray    black   
3  yv1612tb4f1310987    ca       4.1   14282.0  white    black   
4  wba6b2c57ed129731    ca       4.3    2641.0   gray    black   

                                   seller    mmr  sellingprice  \
0                 kia motors america, inc  20500         21500   
1                 kia motors america, inc  20800         21500   
2  financial services remarketing (lease)  31900         30000   
3                 volvo na rep/world omni  27500         27750   
4  financial services remarketing (lease)  66000         67000   

                                  saledate  
0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
1  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)  
3  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)  
4  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)

In [4]:
df.dtypes

year              int64
make             object
model            object
trim             object
body             object
transmission     object
vin              object
state            object
condition        object
odometer        float64
color            object
interior         object
seller           object
mmr              object
sellingprice      int64
saledate         object
dtype: object

Sample narrative description.

* The columns make, model, trim, body, transmission, vin, state, condition, color,interior, seller, mmr, saledate have incorrect type

In [5]:
df=df.convert_dtypes()
df.dtypes

year              Int64
make             string
model            string
trim             string
body             string
transmission     string
vin              string
state            string
condition        string
odometer        Float64
color            string
interior         string
seller           string
mmr              string
sellingprice      Int64
saledate         string
dtype: object

In [6]:
df['condition'] = pd.to_numeric(df['condition'],errors='coerce')
df['mmr'] = pd.to_numeric(df['mmr'],errors='coerce')
df.dtypes

year              Int64
make             string
model            string
trim             string
body             string
transmission     string
vin              string
state            string
condition       float64
odometer        Float64
color            string
interior         string
seller           string
mmr             float64
sellingprice      Int64
saledate         string
dtype: object

## Question 3. Analyze the `make` column.
- Comment on what you observe.  
- Clean the column to reduce the number of unique values to as few as reasonably possible (no hard coding). 

Hint: What do you notice if you run this:
```python
makes = df.make.unique().tolist()
makes = [str(x).lower() for x in makes]
sorted(makes)
```

In [7]:
df.make.sort_values().unique()

<StringArray>
[        'Acura',  'Aston Martin',          'Audi',           'BMW',
       'Bentley',         'Buick',      'Cadillac',     'Chevrolet',
      'Chrysler',        'Daewoo',         'Dodge',          'FIAT',
       'Ferrari',        'Fisker',          'Ford',           'GMC',
           'Geo',        'HUMMER',         'Honda',       'Hyundai',
      'Infiniti',         'Isuzu',        'Jaguar',          'Jeep',
           'Kia',   'Lamborghini',    'Land Rover',         'Lexus',
       'Lincoln',         'Lotus',          'MINI',      'Maserati',
         'Mazda', 'Mercedes-Benz',       'Mercury',    'Mitsubishi',
        'Nissan',    'Oldsmobile',      'Plymouth',       'Pontiac',
       'Porsche',           'Ram',   'Rolls-Royce',          'Saab',
        'Saturn',         'Scion',        'Subaru',        'Suzuki',
         'Tesla',        'Toyota',    'Volkswagen',         'Volvo',
         'acura',     'airstream',          'audi',           'bmw',
         'buick',   

Sample narrative description.

In [8]:
makes = df.make.unique().tolist()
makes = [str(x).lower() for x in makes]
sorted(makes)

['<na>',
 'acura',
 'acura',
 'airstream',
 'aston martin',
 'audi',
 'audi',
 'bentley',
 'bmw',
 'bmw',
 'buick',
 'buick',
 'cadillac',
 'cadillac',
 'chev truck',
 'chevrolet',
 'chevrolet',
 'chrysler',
 'chrysler',
 'daewoo',
 'dodge',
 'dodge',
 'dodge tk',
 'dot',
 'ferrari',
 'fiat',
 'fisker',
 'ford',
 'ford',
 'ford tk',
 'ford truck',
 'geo',
 'gmc',
 'gmc',
 'gmc truck',
 'honda',
 'honda',
 'hummer',
 'hyundai',
 'hyundai',
 'hyundai tk',
 'infiniti',
 'isuzu',
 'jaguar',
 'jeep',
 'jeep',
 'kia',
 'kia',
 'lamborghini',
 'land rover',
 'land rover',
 'landrover',
 'lexus',
 'lexus',
 'lincoln',
 'lincoln',
 'lotus',
 'maserati',
 'maserati',
 'mazda',
 'mazda',
 'mazda tk',
 'mercedes',
 'mercedes-b',
 'mercedes-benz',
 'mercury',
 'mercury',
 'mini',
 'mitsubishi',
 'mitsubishi',
 'nissan',
 'nissan',
 'oldsmobile',
 'oldsmobile',
 'plymouth',
 'plymouth',
 'pontiac',
 'pontiac',
 'porsche',
 'porsche',
 'ram',
 'rolls-royce',
 'saab',
 'saturn',
 'scion',
 'smart',
 '

In [9]:
df['make'] = df['make'].str.strip()
df['make'] = df['make'].str.lower()
makes = df.make.unique().tolist()
makes = [str(x).lower() for x in makes]
sorted(makes)

['<na>',
 'acura',
 'airstream',
 'aston martin',
 'audi',
 'bentley',
 'bmw',
 'buick',
 'cadillac',
 'chev truck',
 'chevrolet',
 'chrysler',
 'daewoo',
 'dodge',
 'dodge tk',
 'dot',
 'ferrari',
 'fiat',
 'fisker',
 'ford',
 'ford tk',
 'ford truck',
 'geo',
 'gmc',
 'gmc truck',
 'honda',
 'hummer',
 'hyundai',
 'hyundai tk',
 'infiniti',
 'isuzu',
 'jaguar',
 'jeep',
 'kia',
 'lamborghini',
 'land rover',
 'landrover',
 'lexus',
 'lincoln',
 'lotus',
 'maserati',
 'mazda',
 'mazda tk',
 'mercedes',
 'mercedes-b',
 'mercedes-benz',
 'mercury',
 'mini',
 'mitsubishi',
 'nissan',
 'oldsmobile',
 'plymouth',
 'pontiac',
 'porsche',
 'ram',
 'rolls-royce',
 'saab',
 'saturn',
 'scion',
 'smart',
 'subaru',
 'suzuki',
 'tesla',
 'toyota',
 'volkswagen',
 'volvo',
 'vw']

Sample narrative description.

* There are duplicates in the 'make' column. I have removed duplicates by eliminating the extra space and lower casing all values.

## Question 4. Analyze the relationship between the wholeprice `mmr` and the `salesprice`.
- What is the overall correlation?  
- What happens to the correlation as you increase the maximum selling price window (e.g., 0-10000, 0-20000,...)?
- Create a scatter plot with the relationship.  
- Create a graph that shows how the correlation changes.

In [10]:
df.head()

year   make                model        trim   body transmission  \
0  2015    kia              Sorento          LX    SUV    automatic   
1  2015    kia              Sorento          LX    SUV    automatic   
2  2014    bmw             3 Series  328i SULEV  Sedan    automatic   
3  2015  volvo                  S60          T5  Sedan    automatic   
4  2014    bmw  6 Series Gran Coupe        650i  Sedan    automatic   

                 vin state  condition  odometer  color interior  \
0  5xyktca69fg566472    ca        5.0   16639.0  white    black   
1  5xyktca69fg561319    ca        5.0    9393.0  white    beige   
2  wba3c1c51ek116351    ca        4.5    1331.0   gray    black   
3  yv1612tb4f1310987    ca        4.1   14282.0  white    black   
4  wba6b2c57ed129731    ca        4.3    2641.0   gray    black   

                                   seller      mmr  sellingprice  \
0                 kia motors america, inc  20500.0         21500   
1                 kia motors america, inc  20800.0         21500   
2  financial services remarketing (lease)  31900.0         30000   
3                 volvo na rep/world omni  27500.0         27750   
4  financial services remarketing (lease)  66000.0         67000   

                                  saledate  
0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
1  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)  
3  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)  
4  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)

In [11]:
df['mmr'].fillna(0, inplace=True)
df['sellingprice'].fillna(0, inplace=True)
df['sellingprice'] = df['sellingprice'].astype(float)

In [12]:
correlation = df['mmr'].corr(df['sellingprice'])
print(correlation)

0.9835870285181985


In [18]:
df['sellingprice'].describe()

count    558837.000000
mean      13611.264979
std        9749.430357
min           1.000000
25%        6900.000000
50%       12100.000000
75%       18200.000000
max      230000.000000
Name: sellingprice, dtype: float64

Sample narrative description.

## Question 5. How does the distribution of `salesprice` vary by make for cars made between 2010-2015?
- Use the 10 most common makes (your cleaned version of this column).  
- Create a graph that answers this question.

In [31]:
df1=df[['year','make','sellingprice']].copy()
df_yearselect = df1[(df1['year'] >= 2010) & (df1['year'] <= 2015)]
df_yearselect

year    make  sellingprice
0       2015     kia       21500.0
1       2015     kia       21500.0
2       2014     bmw       30000.0
3       2015   volvo       27750.0
4       2014     bmw       67000.0
...      ...     ...           ...
558832  2015     kia       33000.0
558833  2012     ram       30800.0
558834  2012     bmw       34000.0
558835  2015  nissan       11100.0
558836  2014    ford       26700.0

[366023 rows x 3 columns]

In [33]:
df_yearselect['make'].value_counts().head(10)

ford         70565
nissan       42374
chevrolet    38161
toyota       26263
dodge        20169
hyundai      17899
honda        15644
kia          15347
infiniti     12720
bmw          11905
Name: make, dtype: Int64

In [41]:
df_dis=df_yearselect.groupby(['year', 'make']).agg({'sellingprice':sum})
df_dis=pd.DataFrame(df_dis)
df_dis=df_dis.reset_index()
df_dis

year        make  sellingprice
0    2010       acura     4208650.0
1    2010        audi     6211350.0
2    2010     bentley      197000.0
3    2010         bmw    11797050.0
4    2010       buick     3848800.0
..    ...         ...           ...
226  2015       smart       27500.0
227  2015      subaru     5791750.0
228  2015      toyota     4574810.0
229  2015  volkswagen     1351450.0
230  2015       volvo    10364750.0

[231 rows x 3 columns]

Sample narrative description.